packages/ai/src/util/async-iterable-stream.ts

```diff
@@ -1,16 +1,87 @@
+/**
+ * A type that combines AsyncIterable and ReadableStream.
+ * This allows a ReadableStream to be consumed using for-await-of syntax.
+ */
 export type AsyncIterableStream<T> = AsyncIterable<T> & ReadableStream<T>;
 
+/**
+ * Wraps a ReadableStream and returns an object that is both a ReadableStream and an AsyncIterable.
+ * This enables consumption of the stream using for-await-of, with proper resource cleanup on early exit or error.
+ *
+ * @template T The type of the stream's chunks.
+ * @param source The source ReadableStream to wrap.
+ * @returns An AsyncIterableStream that can be used as both a ReadableStream and an AsyncIterable.
+ */
 export function createAsyncIterableStream<T>(
   source: ReadableStream<T>,
 ): AsyncIterableStream<T> {
+  // Pipe through a TransformStream to ensure a fresh, unlocked stream.
   const stream = source.pipeThrough(new TransformStream<T, T>());
 
-  (stream as AsyncIterableStream<T>)[Symbol.asyncIterator] = () => {
-    const reader = stream.getReader();
+  /**
+   * Implements the async iterator protocol for the stream.
+   * Ensures proper cleanup (cancelling and releasing the reader) on completion, early exit, or error.
+   */
+  (stream as AsyncIterableStream<T>)[Symbol.asyncIterator] = function (
+    this: ReadableStream<T>,
+  ): AsyncIterator<T> {
+    const reader = this.getReader();
+
+    let finished = false;
+
+    /**
+     * Cleans up the reader by cancelling and releasing the lock.
+     */
+    async function cleanup() {
+      finished = true;
+      try {
+        await reader.cancel?.();
```

Comment:
@remorses this is now also called from the regular `done`, which might be a bug that i might have introduced. would calling cancel on done be okay or would it need to be avoided (in which case i would add another test case for that scenario)?

Reply:
cancel should not be called when the stream completes normally, even if it works in Node there could be other streams implementation like Bun or Deno that fail in this case.